cesql/spec.md

```diff
@@ -210,9 +225,24 @@ returned together with the evaluated value of the CESQL expression.
 Whenever possible, some error checks SHOULD be done at compile time by the expression evaluator, in order to prevent
 runtime errors.
 
+Every CESQL engine MUST support the following error types:
+
+- _ParseError_: An error that occurs during parseing
+- _MathError_: An error that occurs during the evaluation of a mathematical operation
+- _CastError_: An error that occurs during an implicit or explicit type cast
+- _MissingFunctionError_: An error that occurs due to a call to a function that has not been registered with the CESQL engine
+- _FunctionEvaluationError_: An error that occurs during the evaluation of a function
+
+Whenever an operator or function encounters an error, it MUST return a return value as well as an error value or values. In cases where there is not an obvious return value for the expression,
+the operator or function SHOULD return the zero value for the return type of the operator or function.
+
 ### 3.4. Operators
 
-The following tables show the operators that MUST be supported by a CESQL evaluator.
+The following tables show the operators that MUST be supported by a CESQL evaluator. When evaluating an operator,
+a CESQL engine MUST attempt to cast the operands to the correct types. A CESQL evaluator MUST stop evaluating an operator
+when it encounters an error value, and it MUST return both the zero value for the return type of the operator as well
+as the error. In cases where the return type for the operator is ambiguous, the return type MUST be assumed to be
+_Boolean_ and the return value MUST be `false`.
 
 All the operators in the following tables are listed in precedence order.
```

Comment:
Below, it defines "x = y" before "x != y". Given the sentence above, does this mean that:
```
a != b = c
```
must have "b = c" evaluated before the "a != ..." part?

The spec also says "Operators MUST be evaluated in left to right order", which I'd interpret as "a != b" would be evaluated first.

And I'm reading this wrong or are things a bit ambiguous?

Reply:
@duglin I think that what actually happens is that subexpressions are evaluated left to right but that within that there is a precedence (coming from the antlr4 lexer for the language). Technically, what was originally written here is correct about the precedence of operations. Maybe we need to clarify that evaluation is left to right, but that the precedence of operators/functions takes priority over that L to R evaluation?

Reply:
I've removed the text about operators being listed in precedence order, since it did not seem to match the actual precedence from the parser. Instead, I have added a table with the updated precedence.